reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,417 @@
+            this.negatedsCost = -1;
+        }
+
+        public long estimateAllAnswers() {
+            return this.fullAnswerCount;
+        }
+
+        public long estimateAnswers(Set<Variable> projectToVariables, Set<Resolvable<?>> includedResolvables) {
+            List<LocalEstimate> enabledEstimates = iterate(includedResolvables)
+                    .flatMap(resolvable -> iterate(multivarEstimates.get(resolvable)))
+                    .toList();
+            Map<Variable, CostCover> costCover = computeCostCover(enabledEstimates, projectToVariables);
+            long ret = CostCover.costToCover(projectToVariables, costCover);
+            assert ret > 0;             // Flag in tests if it happens.
+            return Math.max(ret, 1);    // Don't do stupid stuff in prod when it happens.
+        }
+
+        private Map<Variable, CostCover> computeCostCover(List<LocalEstimate> enabledEstimates, Set<Variable> projectToVariables) {
+            // Does a greedy set cover
+            Map<Variable, CostCover> costCover = new HashMap<>(unaryCostCover);
+            enabledEstimates.sort(Comparator.comparing(x -> x.answerEstimate(this, projectToVariables)));
+            for (LocalEstimate enabledEstimate : enabledEstimates) {
+                Set<Variable> interestingSubsetOfVariables = enabledEstimate.variables.stream()
+                        .filter(projectToVariables::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = CostCover.costToCover(interestingSubsetOfVariables, costCover);
+                if (currentCostToCover > enabledEstimate.answerEstimate(this, interestingSubsetOfVariables)) {
+                    CostCover variablesNowCoveredBy = new CostCover(enabledEstimate.answerEstimate(this, interestingSubsetOfVariables));
+                    interestingSubsetOfVariables.forEach(v -> costCover.put(v, variablesNowCoveredBy));
+                }
+            }
+            return costCover;
+        }
+
+        private void registerTriggeredRules() {
+            resolvables().filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                iterate(unifiers.keySet()).forEachRemaining(rule -> {
+                    costEstimator.registerConjunction(rule.condition().conjunction());
+                });
+            });
+        }
+
+        private boolean isInferrable(Resolvable<?> resolvable) {
+            return resolvable.isConcludable() &&
+                    !logicMgr.applicableRules(resolvable.asConcludable()).isEmpty();
+        }
+
+        private FunctionalIterator<Resolvable<?>> resolvables() {
+            return iterate(logicMgr.compile(conjunction));
+        }
+
+        private Set<Variable> allVariables() {
+            return iterate(conjunction.pattern().variables())
+                    .filter(v -> v.isThing())
+                    .toSet();
```

Comment:
We don't know how to deal with type variables yet, or whether we need to at all.